File: searching-algorithms/binary-search/binary_search_recursive.cs
/*
Note: This file is designed to be copied out and compiled on your machine. 
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file. 

To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Compile the program
4. Run the program
*/

using System;

namespace isaac_code_samples
{
    class BinarySearchRecursive
    {
        void Main(string[] args){
           test(); 
        }

        public static int BinarySearch(int[] items, int search_item, int first, int last){
            if (first > last){
                return -1;
            }else{
                int midpoint = (first + last) / 2;
                if (items[midpoint] == search_item){
                    return midpoint;
                }else if (search_item > items[midpoint]){
                    first = midpoint + 1;
                    return BinarySearch(items, search_item, first, last);
                }else{
                    last = midpoint - 1;
                    return BinarySearch(items, search_item, first, last);
                }
            }
        }

        private static void test(){
            int[] items = new int[]{5, 12, 24, 56, 68, 72, 81, 95};

            //Search for first item in array
            Console.WriteLine("Searching for first item in array...");
            int result = BinarySearch(items, 5, 0, items.Length - 1);
            Console.WriteLine("The search result was: " + result.ToString());

            //Search for last item in array
            Console.WriteLine("Searching for last item in array...");
            result = BinarySearch(items, 95, 0, items.Length - 1);
            Console.WriteLine("The search result was: " + result.ToString());

            //Search for an item that is not in the array
            Console.WriteLine("Searching for last item in array...");
            result = BinarySearch(items, 36, 0, items.Length - 1);
            Console.WriteLine("The search result was: " + result.ToString());
        }
    }
}

File: searching-algorithms/binary-search/binary_search_recursive.py
def binary_search (items, search_item, first, last):
    """Binary search - recursive version."""
    if first > last:
        return -1
    else:
        midpoint = (first + last) // 2
        if items[midpoint] == search_item:
            return midpoint
        elif search_item > items[midpoint]:
            first = midpoint + 1
            return binary_search(items, search_item, first, last)
        else:
            last = midpoint - 1 
            return binary_search(items, search_item, first, last)

def test():
    """Test script for binary search."""
    # Create a list containing some names
    items = ["Albie", "Byron", "Divya", "Hari", "Joanne", "Kim", "Wilbur", "Zoe"] 
    first = 0
    last = 7
    
    # Search for first item in the list
    print ("Searching for first name in the list")
    result = binary_search(items, "Albie", first, last)
    print ('Result of search was {}'.format(result))
    
    # Search for last item in the list
    print ('Searching for the last name in the list')
    result = binary_search(items, "Zoe", first, last)
    print ('Result of search was {}'.format(result))
    
    # Search for an item that is not in the list
    print ('Searching for a name that is not in the list')
    result = binary_search(items, "Boris", first, last)
    print ('Result of search was {}'.format(result))

if __name__ == '__main__':
    test()
